pkg/cloud/aws/services/userdata/controlplane.go

```diff
@@ -16,111 +16,187 @@ limitations under the License.
-echo -n '{{.EtcdCAKey}}' > /etc/kubernetes/pki/etcd/ca.key
-chmod 600 /etc/kubernetes/pki/etcd/ca.key
-
-echo -n '{{.FrontProxyCACert}}' > /etc/kubernetes/pki/front-proxy-ca.crt
-echo -n '{{.FrontProxyCAKey}}' > /etc/kubernetes/pki/front-proxy-ca.key
-chmod 600 /etc/kubernetes/pki/front-proxy-ca.key
-
-echo -n '{{.SaCert}}' > /etc/kubernetes/pki/sa.pub
-echo -n '{{.SaKey}}' > /etc/kubernetes/pki/sa.key
-chmod 600 /etc/kubernetes/pki/sa.key
-
-PRIVATE_IP=$(curl http://169.254.169.254/latest/meta-data/local-ipv4)
-HOSTNAME="$(curl http://169.254.169.254/latest/meta-data/local-hostname)"
-
-cat >/tmp/kubeadm.yaml <<EOF
----
-apiVersion: kubeadm.k8s.io/v1beta1
-kind: ClusterConfiguration
-apiServer:
-  certSANs:
-    - "$PRIVATE_IP"
-    - "{{.ELBAddress}}"
-  extraArgs:
-    cloud-provider: aws
-controlPlaneEndpoint: "{{.ELBAddress}}:6443"
-clusterName: "{{.ClusterName}}"
-networking:
-  dnsDomain: "{{.ServiceDomain}}"
-  podSubnet: "{{.PodSubnet}}"
-  serviceSubnet: "{{.ServiceSubnet}}"
-kubernetesVersion: "{{.KubernetesVersion}}"
----
-apiVersion: kubeadm.k8s.io/v1beta1
-kind: InitConfiguration
-nodeRegistration:
-  name: ${HOSTNAME}
-  criSocket: /var/run/containerd/containerd.sock
-  kubeletExtraArgs:
-    cloud-provider: aws
-EOF
-
-kubeadm init --config /tmp/kubeadm.yaml --v 10
+	controlPlaneCloudInit = `{{.Header}}
+write_files:
+-   path: /etc/kubernetes/pki/ca.crt
+    encoding: "base64"
+    owner: root:root
+    permissions: '0640'
+    content: |
+      {{.CACert | Base64Encode}}
+
+-   path: /etc/kubernetes/pki/ca.key
+    encoding: "base64"
+    owner: root:root
+    permissions: '0600'
+    content: |
+      {{.CAKey | Base64Encode}}
+
+-   path: /etc/kubernetes/pki/etcd/ca.crt
+    encoding: "base64"
+    owner: root:root
+    permissions: '0640'
+    content: |
+      {{.EtcdCACert | Base64Encode}}
+
+-   path: /etc/kubernetes/pki/etcd/ca.key
+    encoding: "base64"
+    owner: root:root
+    permissions: '0600'
+    content: |
+      {{.EtcdCAKey | Base64Encode}}
+
+-   path: /etc/kubernetes/pki/front-proxy-ca.crt
+    encoding: "base64"
+    owner: root:root
+    permissions: '0640'
+    content: |
+      {{.FrontProxyCACert | Base64Encode}}
+
+-   path: /etc/kubernetes/pki/front-proxy-ca.key
+    encoding: "base64"
+    owner: root:root
+    permissions: '0600'
+    content: |
+      {{.FrontProxyCAKey | Base64Encode}}
+
+-   path: /etc/kubernetes/pki/sa.pub
+    encoding: "base64"
+    owner: root:root
+    permissions: '0640'
+    content: |
+      {{.SaCert | Base64Encode}}
+
+-   path: /etc/kubernetes/pki/sa.key
+    encoding: "base64"
+    owner: root:root
+    permissions: '0600'
+    content: |
+      {{.SaKey | Base64Encode}}
+
+-   path: /tmp/kubeadm.yaml
+    owner: root:root
+    permissions: '0640'
+    content: |
+      ---
+      apiVersion: kubeadm.k8s.io/v1beta1
+      kind: ClusterConfiguration
+      apiServer:
+        certSANs:
+          - {{ "{{ ds.meta_data.local_ipv4 }}" }}
```

Comment:
@sfzylad can you please explain what this syntax means?  :)

Reply:
This value is builtin variable that `cloud-init` uses referring to instance's metadata. I referred to it based on this document: https://cloudinit.readthedocs.io/en/latest/topics/instancedata.html

Reply:
Thank you for the explanation :)